server/server/src/billing/billing_service.rs

```diff
@@ -1,194 +1,361 @@
-                return Err(ClientError(SwitchAccountTierError::CurrentUsageIsMoreThanNewTier));
-            }
+        pipe.json_set(keys::subscription_profile(public_key), &sub_profile)
+    });
+    return_if_error!(tx_result);
 
-            let pos = get_active_subscription_index(&user_info.subscriptions)?;
+    info!(
+        "User successfully entered payment flow. public_key: {}",
+        keys::stringify_public_key(public_key)
+    );
 
-            stripe_client::cancel_subscription(
-                &server_state.stripe_client,
-                &stripe::SubscriptionId::from_str(&user_info.subscriptions[pos].id)?,
-            )
-            .await?;
+    Ok(sub_profile)
+}
+
+async fn reset_billing_lock<E: Debug>(
+    public_key: &PublicKey, con: &mut deadpool_redis::Connection,
+    sub_profile: &mut SubscriptionProfile,
+) -> Result<(), ServerError<E>> {
+    sub_profile.last_in_payment_flow = 0;
+    Ok(con
+        .json_set(keys::subscription_profile(public_key), &sub_profile)
+        .await?)
+}
 
-            info!("Successfully canceled stripe subscription. public_key: {}", fmt_public_key);
+pub async fn get_subscription_info(
+    context: RequestContext<'_, GetSubscriptionInfoRequest>,
+) -> Result<GetSubscriptionInfoResponse, ServerError<GetSubscriptionInfoError>> {
+    let server_state = context.server_state;
+    let mut con = server_state.index_db_pool.get().await?;
 
-            user_info.subscriptions[pos].is_active = false;
+    let sub_profile: SubscriptionProfile = match con
+        .maybe_json_get(keys::subscription_profile(&context.public_key))
+        .await?
+    {
+        Some(sub_profile) => sub_profile,
+        None => return Ok(GetSubscriptionInfoResponse { subscription_info: None }),
+    };
 
-            FREE_TIER_USAGE_SIZE
+    let subscription_info = match (sub_profile.data_cap, sub_profile.billing_platform) {
+        (FREE_TIER_USAGE_SIZE, _) => None,
+        (PREMIUM_TIER_USAGE_SIZE, Some(BillingPlatform::Stripe(info))) => Some(SubscriptionInfo {
+            payment_platform: PaymentPlatform::Stripe { card_last_4_digits: info.last_4.clone() },
+            period_end: info.expiration_time,
+        }),
+        (PREMIUM_TIER_USAGE_SIZE, Some(BillingPlatform::GooglePlay(info))) => {
+            Some(SubscriptionInfo {
+                payment_platform: PaymentPlatform::GooglePlay {
+                    account_state: info.account_state.clone(),
+                },
+                period_end: info.expiration_time,
+            })
         }
-        (_, AccountTier::Free) | (_, AccountTier::Premium(_)) => {
+        (_, _) => {
             return Err(internal!(
-                "Unrecognized current data cap: {}, public_key: {}",
-                current_data_cap,
-                fmt_public_key
-            ));
+                "No billing information despite being a premium tier. public_key: {:?}",
+                context.public_key
+            ))
         }
     };
```

Comment:
I think there's a nicer way to write this:
1. get an opt
2. map the opt
3. return the answer

I think there's more idiomatic ways to express: convert this struct to that struct (number 2)